iree/compiler/Codegen/Common/SetNumWorkgroupsFromLinalgExtPass.cpp

```diff
@@ -60,6 +62,70 @@ class OneToOneRewritePattern : public OpRewritePattern<SourceOp> {
   }
 };
 
+/// Forward LinalgExt::InParallel -> Tensor::InsertSlice -> Flow::TensorStore.
+/// This pattern is necessary for correctness, it accounts for the fact that
+/// InParallel is distributed across multiple workgroups when lowering to HAL
+/// but it then connects to a sequential tensor.insert_slice and then to
+/// flow.dispatch.tensor_store.
+///
+// TODO: All the rewrites in this file this should be done as part of InParallel
+// -> HAL rewrite. But because of dialect dependencies and layering, we have
+// some phase ordering that prevents it atm.
+class ForwardInParallelResultToFlow
```

Comment:
THis is the exact same pattern as https://github.com/google/iree/blob/879f07684af55ca7948b64cc00db460167fd263c/iree/compiler/Codegen/Common/BufferizeCopyOnlyDispatchesPass.cpp#L162 AFAICS. 

THose patterns are going to move to Flow as canonicalizations anyway.  Is there more than `tensor.insert_slice` -> `flow.dispatch.tensor.store` forwarding happening here?

Reply:
Nope, just didn't know better as to look them up..
As I mentioned offline, I may as well adopt your patterns directly.

I was pondering moving them to FlowOps.cpp but that would create a HAL -> Flow dependence that is not there yet.

If you point me to a place where to move that code and expose it I can do that before landing?
I would only create the `populateFoldingXXXPatterns` and move code, not activate as folding/canonicalization in this patch.

Reply:
No need to move it to `Flow`. You can just add the `populateXXX` method to `iree/compiler/Codegen/Utils.h`/`Utils.cpp`. I can move it from there.

Reply:
Upon further inspection, the application logic is not exactly the same, there is still a small impedance gap between inparallel and dispatch ops that will be ironed out over time. 

In the meantime I was able to refactor the core logic a bit (and we should also upstream it).

Reply:
Hmmm, I am not sure what the difference is. If you could file a bug describing the mismatch, that would be great.